### PR TITLE
🍭 Custom reward function for RLOO

### PR DESCRIPTION
**Pull request by August-murr**
Similar to #2540 

Since RLOO doesn't use a value model, it's much simpler.

### CODE REVIEW

File: trl/trainer/utils.py [hunk 1]
```diff
@@ -1198,6 +1198,18 @@ def get_reward(
...
 
+def get_reward_custom(
+    model: Callable, processor: PreTrainedTokenizerBase, query_responses: torch.Tensor
+) -> torch.Tensor:
+    """
+    This function ensures that the custom reward function produces the correct output structure for integration with the trainer script.
+    """
+    texts = processor.batch_decode(query_responses, skip_special_tokens=True)
+    rewards = model(texts)
+    rewards = torch.tensor(rewards, dtype=torch.float).to(query_responses.device)
```
Comment (qgallouedec) [inline trl/trainer/utils.py:1209]: I'm wondering, since it's only three lines and used once, maybe you can put it directly in RLOO?

File: trl/trainer/rloo_trainer.py [hunk 2]
```diff
@@ -79,7 +80,7 @@ def __init__(
         ],
         policy: nn.Module,
         ref_policy: nn.Module,
-        reward_model: nn.Module,
+        reward_model: Union[nn.Module, Callable],
```
Comment (qgallouedec) [inline trl/trainer/rloo_trainer.py:83]:
```suggestion
        reward_model: Union[nn.Module, Callable[[list[str]], list[float]],
```

**Comment by qgallouedec** [comment]
Nice!

Can be done in a follow-up PR, but it might make sense to rename `reward_model` to `reward_func`. See https://github.com/huggingface/trl/pull/2606#discussion_r1925849928 for more context.

File: trl/trainer/utils.py (hunk 1 above)
Comment (August-murr) [reply trl/trainer/utils.py:1209]: but it could be used in other trainers that use `get_reward` like ppo or online dpo

**Comment by August-murr** [comment]
> Nice!
> 
> Can be done in a follow-up PR, but it might make sense to rename `reward_model` to `reward_func`. See [#2606 (comment)](https://github.com/huggingface/trl/pull/2606#discussion_r1925849928) for more context.

yea I think it's better to do it for all trainers in a separate PR

**Review by qgallouedec — CHANGES_REQUESTED** [review]
We just need a unit test and we're good to merge

File: trl/trainer/rloo_trainer.py [hunk 3]
```diff
@@ -350,9 +353,17 @@ def repeat_generator():
...
-                    _, score, _ = get_reward(
-                        reward_model, postprocessed_query_response, processing_class.pad_token_id, context_length
-                    )
+
+                    if isinstance(reward_model, nn.Module):
+                        _, score, _ = get_reward(
+                            reward_model, postprocessed_query_response, processing_class.pad_token_id, context_length
+                        )
+                    else:
+                        score = torch.tensor(
```
Comment (qgallouedec) [inline trl/trainer/rloo_trainer.py:362]: need to add `dtype=torch.float` I think

**Review by qgallouedec — APPROVED** [review]
Nice! Feel free to merge when the CI is green (ignore dev dep that will fail)
